solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkLiveNodes.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.common.cloud;
+
+import static java.util.Collections.emptySortedSet;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.solr.common.SolrException;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ZkLiveNodes implements Watcher {
+  public static final String LIVE_NODES_ZKNODE = "/live_nodes";
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private volatile SortedSet<String> liveNodes = emptySortedSet();
+  private final Set<LiveNodesListener> liveNodesListeners = ConcurrentHashMap.newKeySet();
+  private final SolrZkClient zkClient;
+
+  public ZkLiveNodes(SolrZkClient zkClient) {
+    this.zkClient = zkClient;
+  }
+
+  public SortedSet<String> getLiveNodes() {
+    return liveNodes;
+  }
+
+  protected void setLiveNodes(SortedSet<String> liveNodes) {
+    this.liveNodes = liveNodes;
+  }
+
+  public Set<LiveNodesListener> getLiveNodesListeners() {
+    return liveNodesListeners;
+  }
+
+  public void addLiveNodesListener(LiveNodesListener liveNodesListener) {
+    this.liveNodesListeners.add(liveNodesListener);
+  }
+
+  public void removeLiveNodesListener(LiveNodesListener liveNodesListener) {
+    this.liveNodesListeners.remove(liveNodesListener);
+  }
+
+  /* Refresh live nodes */
+  public synchronized void refreshLiveNodes() throws KeeperException, InterruptedException {
```

Comment:
I think the methods on this class don't need "LiveNodes" in them.

Reply:
true